### PR TITLE
spark backend for sample-counts api

### PR DESCRIPTION
**Pull request by doori**
spark backend for api/sample-counts/fetch.

Reviewers:
@kalletlak @n1zea144

### CODE REVIEW

File: persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java [hunk 1]
```diff
@@ -35,26 +35,76 @@
...
+                df = spark.read()
+                    .format("csv")
+                    .option("delimiter", ":")
+                    .load(inputFile);
+
+                it = df.toLocalIterator();
+                while (it.hasNext()) {
+                    Row row = (Row) it.next();
+                    String colName = row.getString(0);
+                    String listCol = "case".equalsIgnoreCase(typeOfData) ? "case_list_ids" : "gene_list";
```
Comment (n1zea144) [inline persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java]: Does this value ever change?

Comment (doori) [reply persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java]: i see both [case_list_ids](https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#case-lists) and [gene_list](https://docs.cbioportal.org/5.1-data-loading/data-loading/import-gene-panels#gene-panel-file-format) is in the documentation.

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java [hunk 2]
```diff
@@ -36,12 +37,18 @@
     @Qualifier("sampleSparkRepository")
     private SampleRepository sampleSparkRepository;
     @Autowired
+    private CopyNumberSegmentSparkRepository copyNumberSegmentSparkRepository;
```
Comment (n1zea144) [inline service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: Is this not implementing the CopyNumberSegmentRepos?

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java [hunk 3]
```diff
@@ -190,4 +196,30 @@ private void processSamples(List<Sample> samples, String projection) {
             });
         }
     }
+
+    private void processSamplesSpark(List<Sample> samples, String projection) {
```
Comment (n1zea144) [inline service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: Is the only difference here wit processSamples() the call to the spark repos?  If so, suggest passing a ref to the repos into a generalized method.

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java (hunk 2 above)
Comment (doori) [reply service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: This is another case where the method returns internalIds and filters based on that. (`List<Integer> samplesWithCopyNumberSeg = copyNumberSegmentRepository.fetchSamplesWithCopyNumberSegments()`) So the workaround was to create a method that returns `List<String> sampleIds` and create another method in SampleSeriviceImpl.

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java (hunk 3 above)
Comment (doori) [reply service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: That is not the only difference. It's the persistence interface returning internalIds which is something we cannot get from Parquet.

File: persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java (hunk 1 above)
Comment (n1zea144) [reply persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java]: I mean once this routine is invoked, will listCol be assigned the same value each time through the while-loop?

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java (hunk 2 above)
Comment (n1zea144) [reply service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: Why not introduce a method to the interface which returns a List<String> (or return List<Sample> and have the service process internal ids or stable sample ids as required)?

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java (hunk 3 above)
Comment (n1zea144) [reply service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: See comment above.

File: persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java (hunk 1 above)
Comment (doori) [reply persistence/persistence-spark/src/main/java/org/cbioportal/persistence/spark/util/ParquetWriter.java]: Thanks for clarifying. moved it out of the while-loop.

File: service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java (hunk 2 above)
Comment (doori) [reply service/src/main/java/org/cbioportal/service/impl/SampleServiceImpl.java]: thanks for the suggestion. I added a method in the interface that returns a list of Samples.

**Pull request merged**